src/measure_theory/prod.lean

```diff
@@ -0,0 +1,689 @@
+
+/-- A binary function is integrable if the function `x ↦ f (x, y)` is integrable for almost every
+  `y` and the function `y ↦ ∫ ∥f (x, y)∥ dx` is integrable. -/
+lemma integrable_prod_iff' [sigma_finite μ] ⦃f : α × β → E⦄ (h1f : measurable f) :
+  (∀ᵐ y ∂ ν, integrable (λ x, f (x, y)) μ) ∧ integrable (λ y, ∫ x, ∥f (x, y)∥ ∂μ) ν ↔
+  integrable f (μ.prod ν) :=
+by { convert integrable_prod_iff (h1f.comp measurable_swap) using 1, rw [integrable_swap_iff],
+  apply_instance }
+
+lemma integrable.prod_left_ae [sigma_finite μ] ⦃f : α × β → E⦄
+  (hf : integrable f (μ.prod ν)) : ∀ᵐ y ∂ ν, integrable (λ x, f (x, y)) μ :=
+((integrable_prod_iff' hf.measurable).mpr hf).1
+
+lemma integrable.prod_right_ae [sigma_finite μ] ⦃f : α × β → E⦄
+  (hf : integrable f (μ.prod ν)) : ∀ᵐ x ∂ μ, integrable (λ y, f (x, y)) ν :=
+hf.swap.prod_left_ae
+
+lemma integrable.integral_norm_prod_left ⦃f : α × β → E⦄
+  (hf : integrable f (μ.prod ν)) : integrable (λ x, ∫ y, ∥f (x, y)∥ ∂ν) μ :=
+((integrable_prod_iff hf.measurable).mpr hf).2
+
+lemma integrable.integral_norm_prod_right [sigma_finite μ] ⦃f : α × β → E⦄
+  (hf : integrable f (μ.prod ν)) : integrable (λ y, ∫ x, ∥f (x, y)∥ ∂μ) ν :=
+hf.swap.integral_norm_prod_left
+
+end
+
+variables [second_countable_topology E] [normed_space ℝ E]
+  [complete_space E] [borel_space E]
+
+lemma integrable.integral_prod_left ⦃f : α × β → E⦄
+  (hf : integrable f (μ.prod ν)) : integrable (λ x, ∫ y, f (x, y) ∂ν) μ :=
+integrable.mono hf.integral_norm_prod_left hf.measurable.integral_prod_right' $
+  eventually_of_forall $ λ x, (norm_integral_le_integral_norm _).trans_eq $
+  (norm_of_nonneg $ integral_nonneg_of_ae $ eventually_of_forall $ λ y, (norm_nonneg _ : _)).symm
+
+lemma integrable.integral_prod_right [sigma_finite μ] ⦃f : α × β → E⦄
+  (hf : integrable f (μ.prod ν)) : integrable (λ y, ∫ x, f (x, y) ∂μ) ν :=
+hf.swap.integral_prod_left
+
+/-! ### The Bochner integral on a product -/
+
+variables [sigma_finite μ]
+
+lemma integral_prod_swap (f : α × β → E)
+  (hf : measurable f) : ∫ z, f z.swap ∂(ν.prod μ) = ∫ z, f z ∂(μ.prod ν) :=
+by rw [← integral_map measurable_swap hf, prod_swap]
+
+variables {E' : Type*} [measurable_space E'] [normed_group E'] [borel_space E'] [complete_space E']
+  [normed_space ℝ E'] [second_countable_topology E']
+
+/-! Some rules about the sum/difference of double integrals. They follow from `integral_add`, but
+  we separate them out as separate lemmas, because they involve quite some steps. -/
+lemma integral_fn_integral_add ⦃f g : α × β → E⦄
+  {F : E → E'} (hF : measurable F)
+  (hf : integrable f (μ.prod ν))
+  (hg : integrable g (μ.prod ν)) :
+  ∫ x, F (∫ y, f (x, y) + g (x, y) ∂ν) ∂μ = ∫ x, F (∫ y, f (x, y) ∂ν + ∫ y, g (x, y) ∂ν) ∂μ :=
+begin
+  refine integral_congr_ae
+    (hF.comp (hf.add hg).measurable.integral_prod_right')
+    (hF.comp (hf.measurable.integral_prod_right'.add hg.measurable.integral_prod_right')) _,
+  refine hg.prod_right_ae.mp _, refine hf.prod_right_ae.mp _,
```

Comment:
It would be nice to have a version of `filter_upwards` that uses `filter.eventually.mp` and `filter.eventually.mono`. Or make `filter_upwards` recognize if the argument has the form `s ∈ l` or `filter.eventually l p`.

Reply:
Ah, I didn't know the tactic `filter_upwards`. I now changed some proofs to use it. 
After using it, I do get hypotheses and goals of the form `a ∈ {x : α | ... }` (even nested ones), so I have to call `dsimp only [mem_set_of_eq]` to make it actually work.